app/services/TaiService.scala

```diff
@@ -28,53 +29,72 @@ import uk.gov.hmrc.domain.Nino
 
-  val mode:Mode = environment.mode
+  val mode: Mode = environment.mode
   val runModeConfiguration: Configuration = configuration
   lazy val taiUrl = baseUrl("tai")
 
   /**
     * Gets a list of tax components
     */
   def taxComponents(nino: Nino, year: Int)(implicit hc: HeaderCarrier): Future[TaxComponentsResponse] = {
-    withMetricsTimer("get-tax-components") { t =>
-
-      simpleHttp.get[TaxComponentsResponse](s"$taiUrl/tai/$nino/tax-account/$year/tax-components")(
-        onComplete = {
-          case r if r.status >= 200 && r.status < 300 =>
-            t.completeTimerAndIncrementSuccessCounter()
-            TaxComponentsSuccessResponse(TaxComponents.fromJsonTaxComponents(r.json))
+    def call = {
+      withMetricsTimer("get-tax-components") { t =>
+        simpleHttp.get[TaxComponentsResponse](s"$taiUrl/tai/$nino/tax-account/$year/tax-components")(
+          onComplete = {
+            case r if r.status >= 200 && r.status < 300 =>
+              t.completeTimerAndIncrementSuccessCounter()
+              TaxComponentsSuccessResponse(TaxComponents.fromJsonTaxComponents(r.json))
 
-          case r if r.status == NOT_FOUND | r.status == BAD_REQUEST =>
-            t.completeTimerAndIncrementSuccessCounter()
-            Logger.warn("Unable to find tax components from the tai-service")
-            TaxComponentsUnavailableResponse
+            case r if r.status == NOT_FOUND | r.status == BAD_REQUEST => {
+              t.completeTimerAndIncrementSuccessCounter()
+              Logger.warn("Unable to find tax components from the tai-service")
+              TaxComponentsUnavailableResponse
+            }
 
-          case r =>
-            t.completeTimerAndIncrementFailedCounter()
-            Logger.warn(s"Unexpected ${r.status} response getting tax components from the tai-service")
-            TaxComponentsUnexpectedResponse(r)
-        },
-        onError = {
-          case e =>
-            t.completeTimerAndIncrementFailedCounter()
-            Logger.error("Error getting tax components from the tai-service", e)
-            TaxComponentsErrorResponse(e)
+            case r =>
+              t.completeTimerAndIncrementFailedCounter()
+              Logger.warn(s"Unexpected ${r.status} response getting tax components from the tai-service")
+              TaxComponentsUnexpectedResponse(r)
+          },
+          onError = {
+            case e =>
+              t.completeTimerAndIncrementFailedCounter()
+              Logger.error("Error getting tax components from the tai-service", e)
+              TaxComponentsErrorResponse(e)
+          }
+        ).map { r =>
+          Logger.warn(s"Returned response from TAI $r")
+          r
+        }.collect {
+          case response@(TaxComponentsSuccessResponse(_) | TaxComponentsUnavailableResponse) => response
         }
-      )
+      }
     }
+    circuitBreaker.withCircuitBreaker(call).fallbackTo(Future.successful(TaiNotAvailable))
```

Comment:
instead of implementing the circuit breaker direct on the service it may be better to use named dependencies and implement it as it's own implementation of the `TaiService` component. This will help break up the concerns and also make switching between the cb pattern and no cb pattern simple

As an example
```
trait TaiService {
  def taxComponents(nino: Nino, year: Int)(implicit hc: HeaderCarrier): Future[TaxComponentsResponse]
}

@Singleton
class CircuitBreakerTaiService @Inject()(@Named("default") underlying: TaiService, circuitBreaker) extends TaiService {
  def taxComponents(nino: Nino, year: Int)(implicit hc: HeaderCarrier): Future[TaxComponentsResponse] = {
    val call = userlying.taxComponents(nino, year).collect {
       case response@(TaxComponentsSuccessResponse(_) | TaxComponentsUnavailableResponse) => response
    }

    circuitBreaker.withCircuitBreaker(call).fallbackTo(Future.successful(TaiNotAvailable))
  }
}

class DefaultTaiService { ... remains the same... }
```
and also add to `LocalGuiceModule`
```
bind[TaiService].qualifiedWith("default").to[DefaultTaiService]
```